File: mandalay_bay/sport_simulator.py
from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any

from blackjack.rng import SECURE_RANDOM

_CATALOG: dict[str, Any] | None = None


def catalog_path() -> Path:
    return Path(__file__).resolve().parent / "data" / "sports_catalog.json"


def load_catalog() -> dict[str, Any]:
    global _CATALOG
    if _CATALOG is not None:
        return _CATALOG
    with catalog_path().open(encoding="utf-8") as f:
        _CATALOG = json.load(f)
    return _CATALOG


def get_sport_keys(catalog: dict[str, Any]) -> list[str]:
    return list(catalog["sports"].keys())


def _participants(sport_def: dict[str, Any]) -> list[dict[str, Any]]:
    return sport_def.get("teams") or sport_def.get("fighters") or sport_def.get("players") or []


def _short_name(full_name: str) -> str:
    return full_name.split()[-1]


def _gaussian_int(mean: float, std: float, min_val: int, max_val: int) -> int:
    total = sum(SECURE_RANDOM.random() for _ in range(6))
    z = (total - 3) / math.sqrt(0.5)
    val = round(mean + z * std)
    return max(min_val, min(max_val, val))


def _rating_diff_to_american_odds(diff: float) -> tuple[int, int]:
    if diff >= 80:
        return -200, 170
    if diff >= 40:
        return -150, 130
    if diff >= 15:
        return -130, 110
    if diff <= -80:
        return 170, -200
    if diff <= -40:
        return 130, -150
    if diff <= -15:
        return 110, -130
    return -110, -110


def _round_spread(value: float, step: float) -> float:
    if not step:
        return 0.0
    return round(value / step) * step


def _pick_matchup(sport_def: dict[str, Any], prefer_division: bool = True) -> tuple[dict[str, Any], dict[str, Any]]:
    pool = _participants(sport_def)[:]
    SECURE_RANDOM.shuffle(pool)
    if len(pool) < 2:
        return pool[0], pool[1]

    if prefer_division and SECURE_RANDOM.random() < 0.7:
        by_div: dict[str, list[dict[str, Any]]] = {}
        for p in pool:
            key = p.get("division") or p.get("weightClass") or p.get("tour") or "default"
            by_div.setdefault(key, []).append(p)
        viable = [g for g in by_div.values() if len(g) >= 2]
        if viable:
            SECURE_RANDOM.shuffle(viable)
            group = viable[0][:]
            SECURE_RANDOM.shuffle(group)
            return group[0], group[1]

    return pool[0], pool[1]


def generate_event(catalog: dict[str, Any], sport_key: str) -> dict[str, Any]:
    sport_def = catalog["sports"][sport_key]
    scoring_type = sport_def.get("scoringProfile", {}).get("type", "points")
    is_outright = scoring_type == "outright"

    if is_outright:
        pool = _participants(sport_def)[:]
        SECURE_RANDOM.shuffle(pool)
        field = pool[: min(4, len(pool))]
        favorite = max(field, key=lambda p: p["powerRating"])
        odds_map: dict[str, int] = {}
        for p in field:
            gap = favorite["powerRating"] - p["powerRating"]
            if gap == 0:
                odds_map[p["name"]] = sport_def.get("juice", 200)
            elif gap <= 20:
                odds_map[p["name"]] = 350
            elif gap <= 40:
                odds_map[p["name"]] = 600
            else:
                odds_map[p["name"]] = 900

        home = field[0]["name"]
        away = field[1]["name"] if len(field) > 1 else field[0]["name"]
        return {
            "eventId": f"{sport_key}-{SECURE_RANDOM.randrange(10000, 99999)}",
            "sport": sport_key,
            "sportLabel": sport_def["label"],
            "eventType": "outright",
            "home": home,
            "away": away,
            "field": [p["name"] for p in field],
            "outrightOdds": odds_map,
            "homeOdds": odds_map.get(home, 200),
            "awayOdds": odds_map.get(away, 350),
            "spread": 0.0,
            "total": 0.0,
            "spreadHomeOdds": sport_def.get("juice", -110),
            "spreadAwayOdds": sport_def.get("juice", -110),
            "totalOverOdds": -110,
            "totalUnderOdds": -110,
            "props": [],
            "homeScore": 0,
            "awayScore": 0,
            "winner": None,
            "propOutcomes": {},
            "status": "scheduled",
            "settled": False,
            "live": False,
            "label": f"{_short_name(away)} vs {_short_name(home)}",
        }

    away_team, home_team = _pick_matchup(sport_def)
    home_rating = home_team["powerRating"] + (sport_def.get("homeField") or 0) * 10
    away_rating = away_team["powerRating"]
    diff = home_rating - away_rating
    home_odds, away_odds = _rating_diff_to_american_odds(diff)
    spread = _round_spread((diff / 25) - (sport_def.get("homeField") or 0), sport_def.get("spreadStep") or 0.5)
    profile = sport_def["scoringProfile"]
    total = round((profile.get("meanTotal") or 44) * 2) / 2

    props: list[dict[str, Any]] = [
        {"id": "both-score", "label": "Both teams score", "yesOdds": -130, "noOdds": 110},
    ]
    if sport_key in ("NFL", "NCAAF"):
        props.append({"id": "over-tds", "label": "Combined TDs over 4.5", "yesOdds": -105, "noOdds": -115})

    return {
        "eventId": (
            f"{sport_key}-{_short_name(home_team['name'])}-"
            f"{_short_name(away_team['name'])}-{SECURE_RANDOM.randrange(1000, 9999)}"
        ),
        "sport": sport_key,
        "sportLabel": sport_def["label"],
        "eventType": "game",
        "home": home_team["name"],
        "away": away_team["name"],
        "homeOdds": home_odds,
        "awayOdds": away_odds,
        "spread": spread,
        "total": total,
        "spreadHomeOdds": sport_def.get("juice", -110),
        "spreadAwayOdds": sport_def.get("juice", -110),
        "totalOverOdds": -110,
        "totalUnderOdds": -110,
        "props": props,
        "homeScore": 0,
        "awayScore": 0,
        "winner": None,
        "propOutcomes": {},
        "status": "scheduled",
        "settled": False,
        "live": False,
        "label": f"{_short_name(away_team['name'])} @ {_short_name(home_team['name'])}",
    }


def generate_board(catalog: dict[str, Any], count: int | None = None) -> list[dict[str, Any]]:
    board_size = count or catalog.get("boardSize", 10)
    sport_keys = get_sport_keys(catalog)
    events: list[dict[str, Any]] = []
    used: set[str] = set()

    for _ in range(board_size):
        sport_key = sport_keys[SECURE_RANDOM.randrange(0, len(sport_keys))]
        attempts = 0
        while True:
            event = generate_event(catalog, sport_key)
            attempts += 1
            key = f"{event['sport']}:{event['home']}:{event['away']}"
            if event.get("eventType") != "game" or key not in used or attempts >= 8:
                if event.get("eventType") == "game":
                    used.add(key)
                break
        events.append(event)
    return events


def _simulate_game_score(catalog: dict[str, Any], event: dict[str, Any]) -> None:
    sport_def = catalog["sports"][event["sport"]]
    profile = sport_def["scoringProfile"]
    participants = {p["name"]: p for p in _participants(sport_def)}
    home_rating = participants.get(event["home"], {}).get("powerRating", 1450)
    away_rating = participants.get(event["away"], {}).get("powerRating", 1450)
    rating_diff = home_rating - away_rating + (sport_def.get("homeField") or 0) * 10

    mean_home = (profile.get("meanTotal") or 44) / 2 + rating_diff / 40
    mean_away = (profile.get("meanTotal") or 44) / 2 - rating_diff / 40
    std = profile.get("stdPerTeam") or 7

    home_score = _gaussian_int(mean_home, std, profile.get("minPerTeam", 0), profile.get("maxPerTeam", 50))
    away_score = _gaussian_int(mean_away, std, profile.get("minPerTeam", 0), profile.get("maxPerTeam", 50))

    if profile.get("type") in ("runs", "goals"):
        home_score = max(0, home_score)
        away_score = max(0, away_score)

    if profile.get("allowTie") and SECURE_RANDOM.random() < 0.22:
        draw = min(home_score, away_score)
        home_score = away_score = draw
    elif not profile.get("allowTie") and home_score == away_score:
        if rating_diff >= 0:
            home_score += 1
        else:
            away_score += 1

    event["homeScore"] = home_score
    event["awayScore"] = away_score
    event["winner"] = event["home"] if home_score > away_score else event["away"] if away_score > home_score else None
    event["propOutcomes"] = {
        "both-score": home_score > 0 and away_score > 0,
        "over-tds": home_score + away_score > 45,
    }


def _simulate_outright(catalog: dict[str, Any], event: dict[str, Any]) -> None:
    sport_def = catalog["sports"][event["sport"]]
    field_names = event.get("field") or [event["home"], event["away"]]
    pool = [p for p in _participants(sport_def) if p["name"] in field_names]
    weights = [max(1, p["powerRating"] - 1300) for p in pool]
    total = sum(weights)
    roll = SECURE_RANDOM.random() * total
    winner = pool[0]["name"]
    for i, w in enumerate(weights):
        roll -= w
        if roll <= 0:
            winner = pool[i]["name"]
            break
    event["winner"] = winner
    if len(field_names) == 2:
        event["homeScore"] = 1 if winner == event["home"] else 0
        event["awayScore"] = 1 if winner == event["away"] else 0


def simulate_event_outcome(catalog: dict[str, Any], event: dict[str, Any]) -> dict[str, Any]:
    sport_def = catalog["sports"][event["sport"]]
    if sport_def.get("scoringProfile", {}).get("type") == "outright":
        _simulate_outright(catalog, event)
    else:
        _simulate_game_score(catalog, event)
    event["status"] = "final"
    event["settled"] = True
    return event

File: docs/js/sportSimulator.js
import { secureRandomInt } from "./core.js";

const CATALOG_PATH = new URL("../data/sports_catalog.json", import.meta.url).href;
const SCENARIOS_PATH = new URL("../data/sports_scenarios.json", import.meta.url).href;
let catalogCache = null;
let scenariosCache = null;

export async function loadCatalog() {
  if (catalogCache) return catalogCache;
  const res = await fetch(CATALOG_PATH);
  if (!res.ok) throw new Error(`Failed to load sports catalog: ${res.status}`);
  catalogCache = await res.json();
  return catalogCache;
}

export function loadCatalogSync(catalog) {
  catalogCache = catalog;
  return catalog;
}

export async function loadSportsScenarios() {
  if (scenariosCache) return scenariosCache;
  const res = await fetch(SCENARIOS_PATH);
  if (!res.ok) throw new Error(`Failed to load sports scenarios: ${res.status}`);
  scenariosCache = await res.json();
  return scenariosCache;
}

export function loadSportsScenariosSync(data) {
  scenariosCache = data;
  return data;
}

/** Materialize a stored scenario into a live board event. */
export function eventFromScenario(scenario, index = 0) {
  return {
    ...scenario,
    eventId: `${scenario.scenarioId}-${index}-${secureRandomInt(1000, 9999)}`,
    homeScore: 0,
    awayScore: 0,
    winner: null,
    propOutcomes: {},
    status: "scheduled",
    settled: false,
    live: false,
    label: scenario.label,
  };
}

/**
 * Page through the stored scenario DB (wraps). Falls back to [] if empty.
 * @param {{ includeSalonOnly?: boolean }} [opts] — main floor excludes salonOnly rows.
 */
export function boardFromScenarios(scenarioDb, cursor = 0, count = null, opts = {}) {
  const raw = scenarioDb?.scenarios ?? [];
  const scenarios = opts.includeSalonOnly
    ? raw
    : raw.filter((s) => !s.salonOnly);
  if (!scenarios.length) return { events: [], nextCursor: 0 };
  const boardSize = count ?? scenarioDb.boardSize ?? 10;
  const events = [];
  let idx = ((cursor % scenarios.length) + scenarios.length) % scenarios.length;
  for (let i = 0; i < boardSize; i += 1) {
    events.push(eventFromScenario(scenarios[idx], i));
    idx = (idx + 1) % scenarios.length;
  }
  return { events, nextCursor: idx };
}

export function getSportKeys(catalog) {
  return Object.keys(catalog.sports);
}

function participantsForSport(sportDef) {
  return sportDef.teams ?? sportDef.fighters ?? sportDef.players ?? [];
}

function shortName(fullName) {
  const parts = fullName.split(" ");
  return parts[parts.length - 1];
}

function gaussianInt(rng, mean, std, minVal, maxVal) {
  let sum = 0;
  for (let i = 0; i < 6; i += 1) sum += rng();
  const z = (sum - 3) / Math.sqrt(0.5);
  const val = Math.round(mean + z * std);
  return Math.max(minVal, Math.min(maxVal, val));
}

function ratingDiffToAmericanOdds(diff) {
  if (diff >= 80) return { home: -200, away: 170 };
  if (diff >= 40) return { home: -150, away: 130 };
  if (diff >= 15) return { home: -130, away: 110 };
  if (diff <= -80) return { home: 170, away: -200 };
  if (diff <= -40) return { home: 130, away: -150 };
  if (diff <= -15) return { home: 110, away: -130 };
  return { home: -110, away: -110 };
}

function roundSpread(value, step) {
  if (!step) return 0;
  return Math.round(value / step) * step;
}

function shuffleInPlace(arr, rngFn = secureRandomInt) {
  for (let i = arr.length - 1; i > 0; i -= 1) {
    const j = typeof rngFn === "function" && rngFn.length >= 2
      ? rngFn(0, i)
      : secureRandomInt(0, i);
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

function pickMatchupParticipants(sportDef, rng, preferDivision = true) {
  const pool = [...participantsForSport(sportDef)];
  shuffleInPlace(pool, secureRandomInt);
  if (pool.length < 2) return [pool[0], pool[1]];

  if (preferDivision && rng() < 0.7) {
    const byDivision = {};
    for (const p of pool) {
      const key = p.division ?? p.weightClass ?? p.tour ?? "default";
      if (!byDivision[key]) byDivision[key] = [];
      byDivision[key].push(p);
    }
    const viable = Object.values(byDivision).filter((g) => g.length >= 2);
    if (viable.length) {
      shuffleInPlace(viable, secureRandomInt);
      const group = [...viable[0]];
      shuffleInPlace(group, secureRandomInt);
      return [group[0], group[1]];
    }
  }

  return [pool[0], pool[1]];
}

function makeRng(seedFn) {
  return () => {
    const n = seedFn(0, 999999);
    return n / 999999;
  };
}

export function generateEvent(catalog, sportKey, options = {}) {
  const rng = options.rng ?? makeRng(secureRandomInt);
  const sportDef = catalog.sports[sportKey];
  if (!sportDef) throw new Error(`Unknown sport: ${sportKey}`);

  const scoringType = sportDef.scoringProfile?.type ?? "points";
  const isOutright = scoringType === "outright";

  if (isOutright) {
    const pool = [...participantsForSport(sportDef)];
    shuffleInPlace(pool, secureRandomInt);
    const field = pool.slice(0, Math.min(4, pool.length));
    const favorite = field.reduce((a, b) => (a.powerRating >= b.powerRating ? a : b));
    const oddsMap = {};
    for (const p of field) {
      const gap = favorite.powerRating - p.powerRating;
      if (gap === 0) oddsMap[p.name] = sportDef.juice ?? 200;
      else if (gap <= 20) oddsMap[p.name] = 350;
      else if (gap <= 40) oddsMap[p.name] = 600;
      else oddsMap[p.name] = 900;
    }
    const eventId = `${sportKey}-${secureRandomInt(10000, 99999)}`;
    const home = field[0].name;
    const away = field[1]?.name ?? field[0].name;

    return {
      eventId,
      sport: sportKey,
      sportLabel: sportDef.label,
      eventType: "outright",
      home,
      away,
      field: field.map((p) => p.name),
      outrightOdds: oddsMap,
      homeOdds: oddsMap[home] ?? 200,
      awayOdds: oddsMap[away] ?? 350,
      spread: 0,
      total: 0,
      spreadHomeOdds: sportDef.juice ?? -110,
      spreadAwayOdds: sportDef.juice ?? -110,
      totalOverOdds: -110,
      totalUnderOdds: -110,
      props: [],
      homeScore: 0,
      awayScore: 0,
      winner: null,
      status: "scheduled",
      settled: false,
      live: false,
      label: `${shortName(away)} vs ${shortName(home)}`,
    };
  }

  const [awayTeam, homeTeam] = pickMatchupParticipants(sportDef, rng);
  const homeRating = homeTeam.powerRating + (sportDef.homeField ?? 0) * 10;
  const awayRating = awayTeam.powerRating;
  const diff = homeRating - awayRating;
  const odds = ratingDiffToAmericanOdds(diff);
  const spread = roundSpread((diff / 25) - (sportDef.homeField ?? 0), sportDef.spreadStep ?? 0.5);
  const profile = sportDef.scoringProfile;
  const total = Math.round((profile.meanTotal ?? 44) * 2) / 2;

  const props = [
    { id: "both-score", label: "Both teams score", yesOdds: -130, noOdds: 110 },
  ];
  if (sportKey === "NFL" || sportKey === "NCAAF") {
    props.push({ id: "over-tds", label: "Combined TDs over 4.5", yesOdds: -105, noOdds: -115 });
  }

  const eventId = `${sportKey}-${shortName(homeTeam.name)}-${shortName(awayTeam.name)}-${secureRandomInt(1000, 9999)}`;

  return {
    eventId,
    sport: sportKey,
    sportLabel: sportDef.label,
    eventType: "game",
    home: homeTeam.name,
    away: awayTeam.name,
    homeOdds: odds.home,
    awayOdds: odds.away,
    spread,
    total,
    spreadHomeOdds: sportDef.juice ?? -110,
    spreadAwayOdds: sportDef.juice ?? -110,
    totalOverOdds: -110,
    totalUnderOdds: -110,
    props,
    homeScore: 0,
    awayScore: 0,
    winner: null,
    propOutcomes: {},
    status: "scheduled",
    settled: false,
    live: false,
    label: `${shortName(awayTeam.name)} @ ${shortName(homeTeam.name)}`,
  };
}

export function generateBoard(catalog, count = null, options = {}) {
  const boardSize = count ?? catalog.boardSize ?? 10;
  const sportKeys = getSportKeys(catalog);
  const events = [];
  const usedMatchups = new Set();

  for (let i = 0; i < boardSize; i += 1) {
    const sportKey = sportKeys[secureRandomInt(0, sportKeys.length - 1)];
    let event;
    let attempts = 0;
    do {
      event = generateEvent(catalog, sportKey, options);
      attempts += 1;
    } while (
      attempts < 8
      && event.eventType === "game"
      && usedMatchups.has(`${event.sport}:${event.home}:${event.away}`)
    );
    if (event.eventType === "game") {
      usedMatchups.add(`${event.sport}:${event.home}:${event.away}`);
    }
    events.push(event);
  }
  return events;
}

function simulateGameScore(catalog, event, rng) {
  const sportDef = catalog.sports[event.sport];
  const profile = sportDef.scoringProfile;
  const participants = participantsForSport(sportDef);
  const homeRating = participants.find((t) => t.name === event.home)?.powerRating ?? 1450;
  const awayRating = participants.find((t) => t.name === event.away)?.powerRating ?? 1450;
  const ratingDiff = homeRating - awayRating + (sportDef.homeField ?? 0) * 10;

  const meanHome = (profile.meanTotal ?? 44) / 2 + ratingDiff / 40;
  const meanAway = (profile.meanTotal ?? 44) / 2 - ratingDiff / 40;
  const std = profile.stdPerTeam ?? 7;

  let homeScore = gaussianInt(rng, meanHome, std, profile.minPerTeam ?? 0, profile.maxPerTeam ?? 50);
  let awayScore = gaussianInt(rng, meanAway, std, profile.minPerTeam ?? 0, profile.maxPerTeam ?? 50);

  if (profile.type === "runs" || profile.type === "goals") {
    homeScore = Math.max(0, homeScore);
    awayScore = Math.max(0, awayScore);
  }

  if (profile.allowTie && rng() < 0.22) {
    const drawScore = Math.min(homeScore, awayScore);
    homeScore = drawScore;
    awayScore = drawScore;
  } else if (!profile.allowTie && homeScore === awayScore) {
    if (ratingDiff >= 0) homeScore += 1;
    else awayScore += 1;
  }

  event.homeScore = homeScore;
  event.awayScore = awayScore;
  event.winner = homeScore > awayScore ? event.home : awayScore > homeScore ? event.away : null;
  event.propOutcomes = {
    "both-score": homeScore > 0 && awayScore > 0,
    "over-tds": homeScore + awayScore > 45,
  };
}

function simulateOutright(catalog, event, rng) {
  const sportDef = catalog.sports[event.sport];
  const pool = participantsForSport(sportDef).filter((p) => (event.field?.includes(p.name) ?? true));
  const weights = pool.map((p) => Math.max(1, p.powerRating - 1300));
  const total = weights.reduce((a, b) => a + b, 0);
  let roll = rng() * total;
  let winner = pool[0].name;
  for (let i = 0; i < pool.length; i += 1) {
    roll -= weights[i];
    if (roll <= 0) {
      winner = pool[i].name;
      break;
    }
  }
  event.winner = winner;
  if (event.field?.length === 2) {
    event.homeScore = event.winner === event.home ? 1 : 0;
    event.awayScore = event.winner === event.away ? 1 : 0;
  }
}

export function simulateEventOutcome(catalog, event, options = {}) {
  if (!catalog) return event;
  const rng = options.rng ?? makeRng(secureRandomInt);
  const sportDef = catalog.sports[event.sport];
  if (!sportDef) return event;

  if (event.eventType === "outright" || event.eventType === "futures"
      || sportDef.scoringProfile?.type === "outright") {
    simulateOutright(catalog, event, rng);
  } else {
    simulateGameScore(catalog, event, rng);
    // Prop outcomes for richer stored props
    const props = event.props ?? [];
    event.propOutcomes = event.propOutcomes ?? {};
    for (const prop of props) {
      if (prop.id === "both-score") {
        event.propOutcomes[prop.id] = event.homeScore > 0 && event.awayScore > 0;
      } else if (prop.id === "over-tds") {
        event.propOutcomes[prop.id] = (event.homeScore + event.awayScore) > 45;
      } else if (prop.id === "over-threes") {
        event.propOutcomes[prop.id] = (event.homeScore + event.awayScore) > 200;
      } else if (prop.id === "over-hrs") {
        event.propOutcomes[prop.id] = (event.homeScore + event.awayScore) > 8;
      } else {
        event.propOutcomes[prop.id] = rng() < 0.5;
      }
    }
  }
  event.status = "final";
  event.settled = true;
  return event;
}

export function applyLiveFixture(event, fixture) {
  const updated = { ...event, ...fixture, live: true };
  if (fixture.homeScore != null) updated.homeScore = fixture.homeScore;
  if (fixture.awayScore != null) updated.awayScore = fixture.awayScore;
  if (fixture.status) updated.status = fixture.status;
  if (fixture.homeScore != null && fixture.awayScore != null) {
    updated.winner = fixture.homeScore > fixture.awayScore
      ? updated.home
      : fixture.awayScore > fixture.homeScore
        ? updated.away
        : null;
    updated.settled = fixture.status === "final";
  }
  return updated;
}
